src/graphql.tsx

```diff
@@ -577,7 +577,7 @@ export default function graphql(
         const clientProps = this.calculateResultProps(data);
         const mergedPropsAndData = assign({}, props, clientProps);
 
-        if (!shouldRerender && renderedElement) {
+        if (!shouldRerender && renderedElement && renderedElement.type === WrappedComponent) {
```

Comment:
I find it hard to believe that `WrappedComponent` would ever referentially change without a new instance of the `GraphQL` component being created 😣

Unless React Hot Loader copies over all of the properties on the old component to the new component, so a `renderedElement` with the incorrect `type` may be copied over. Can you confirm this happens?

More importantly, can you confirm that this actually fixes your issue?

If so, this looks fine to me. If we could just get a comment that reads something like this 😊:

```js
// If we were not marked to perform a new render, and we have a previous
// rendered element of the same type as the component we are wrapping then let
// us just return that previous element instead of rendering a new one.
//
// We check the type because it may be possible with some React hot loading
// implementations that a `renderedElement` is copied over with the old
// component type. Since we do not want to render the old element again, the
// type check guarantees we render a new element in this case.
```